src/EventStore.Projections.Core/Services/Interpreted/JintProjectionStateHandler.cs

```diff
@@ -993,9 +987,266 @@ private bool EnsureLinkMetadata(out JsValue value)
 
 			public void HandleDeleted(JsValue state, string partition, bool isSoftDelete) {
 				if (_deleted != null) {
-					_deleted.Call(this, new JsValue[] {state, Null, partition, isSoftDelete });
+					_deleted.Call(this, new JsValue[] { state, Null, partition, isSoftDelete });
+				}
+			}
+
+			private readonly Serializer _serializer = new Serializer();
+			public string Serialize(JsValue value) {
+				return JintProjectionStateHandler.Serialize(_serializer, value);
+			}
+		}
+
+		public static string Serialize(JsValue value) {
+			return Serialize(new Serializer(), value);
+		}
+
+		private static string Serialize(Serializer serializer, JsValue instance) {
+			var serialized = serializer.Serialize(instance);
+			return Encoding.UTF8.GetString(serialized.Span);
+		}
+
+		private class Serializer {
+			private readonly IWriter[] _iterators;
+			private readonly ArrayBufferWriter<byte> _bufferWriter;
+			private Utf8JsonWriter _writer;
+			private Dictionary<string, JsonEncodedText> _knownPropertyNames;
+			private int _depth;
+			private static readonly IWriter _completed = new CompletedWriter();
+			public Serializer() {
+				_iterators = new IWriter[64];
+				_bufferWriter = new ArrayBufferWriter<byte>(1024 * 1024);
+				_writer = new Utf8JsonWriter(_bufferWriter, new JsonWriterOptions { Indented = false, SkipValidation = true });
+				_knownPropertyNames = new Dictionary<string, JsonEncodedText>();
+			}
+
+			public ReadOnlyMemory<byte> Serialize(JsValue value) {
+				_depth = -1;
+				_bufferWriter.Clear();
+				_writer.Reset();
+
+				if (value is ArrayInstance ai) {
+					PushWriter(new ArrayWriter(ai, this));
+				} else if (value is ObjectInstance oi) {
+					PushWriter(new ObjectWriter(oi, this));
+				} else {
+					PushWriter(new PrimitiveWriter(value));
+				}
+				while (_iterators[_depth].Write(_writer)) { }
+				_writer.Flush();
+				Debug.Assert(_depth == -1);
```

Comment:
im adding some other unit tests and this assertion is failing when serialising primitives, depth is being left as `0`, perhaps because the primitivewriter doesn't pop itself

Reply:
Assertion is no longer valid